doc/feature/synapses.md

```diff
@@ -0,0 +1,240 @@
+    const float* postSurfacePositionX() const;
+    const float* postSurfacePositionY() const;
+    const float* postSurfacePositionZ() const;
+    const float* postCenterPositionX() const;
+    const float* postCenterPositionY() const;
+    const float* postCenterPositionZ() const;
+
+    const size_t* indices() const;
+    const float* conductance() const;
+    const float* utilization() const;
+    const float* depression() const;
+    const float* facilitation() const;
+    const float* decay() const;
+    const int* efficacy() const;
+};
+
+// sequentially & forwarding iteration through synapses for any number of GIDs
+// not thread-safe, but reentrant
+class SynapsesStream
+{
+    bool eos() const;
+
+    // Remaining count reads before eos()
+    // equals to max - current, max determined by get..Synapses() in
+    // brain::Circuit
+    size_t getRemaining() const;
+
+    // - async load of synapses for count cells
+    // - no order guarantee
+    // - resulting Synapses can be empty
+    // - count will be clamped if count > getRemaining()
+    std::future< Synapses > read( size_t count = 1 ) const;
+};
+
+```
+
+## File format
+
+Uses existing nrn* files
+
+## Examples
+
+### Stream synapse positions for pathway density voxelization
+
+```cpp
+const brain::Circuit circuit;
+const brain::SynapsesStream& stream =
+   circuit.getProjectedSynapses( circuit.getGIDs( "mtypetarget_1" ),
+                                 circuit.getGIDs( "mtypetarget_2" ));
+std::future< brain::Synapses > future = stream.read();
+while( !stream.eos( ))
+{
+    const brain::Synapses synapses = future.get();
+    future = stream.read(); // fetch next
+
+    const float* __restrict__ posx = synapses.preSurfacePositionX();
+    const float* __restrict__ posy = synapses.preSurfacePositionY();
+    const float* __restrict__ posz = synapses.preSurfacePositionZ();
+    for( size_t i = 0; i < synapses.size(); ++i )
+        doVoxelization( posx[i], posy[i], posz[i] );
+}
+```
+
+### Access all afferent synapses of a GIDSet
+
+```cpp
+const brain::Circuit circuit;
+const brain::Synapses& synapses =
+    circuit.getAfferentSynapses( circuit.getGIDs( "Layer1" ));
+for( const auto& synapse : synapses )
+    std::cout << synapse.postGID() << " " << synapse.decay() << std::endl;
+```
+
+### Retrograde projection
+
+```cpp
+const brain::Circuit circuit;
+const brion::GIDSet& preNeurons = circuit.getGIDs( "Layer1" );
+const brion::GIDSet postNeuron = { 1 };
+const brain::Synapses& synapses =
+    circuit.getProjectedSynapses( preNeurons, postNeuron );
+BOOST_CHECK( !synapses.empty( ));
+BOOST_CHECK_EQUAL( synapses.size(), 5 );
+```
+
+## Implementation
+
+* Data owned by container in the form of the arrays (transformed from
+  SynapseMatrix), loaded upon construction wrt prefetch, lazy on first touch
+  except for pre- and postGID for connectivity queries and to fulfill the
+  contract for size().
+* Synapse object stores index to get data from (parenting) container.
```

Comment:
Don't forget in the implementation that if a Synapse object touches the dataset it has also to perform the lazy loading. Maybe it's worth commenting it here.